src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -68,80 +77,286 @@ import           Database.LSMTree.Internal.Serialise
+  is no longer true in the presence of collisions. Let's illustrate this using
+  an example of two contiguous pages @pi@ and @pj@:
+
+  > minKey pi == "00000000"
+  > maxKey pi == "10000000"
+  > minKey pj == "10000001"
+  > maxKey pj == "11111111"
+
+  Say we store only the 4 most significant bits (left-to-right) in our compact
+  index. This means that those 4 bits of @maxKey pi@ and @minKey pj@ collide.
+  The problem is that the interval @[minKey pi, maxKey pi]@ strictly contains
+  the interval @[minKey pi, minKey pj)@! The first 4 bits of @minKey pj@ do not
+  actually define the boundary between @pi@ and @pj@, and so there would be no
+  way to answer if a search key with first 4 bits "1000" could be found in @pi@
+  or @pj@. We call such a situation a /clash/,
+
+  The solution in these cases is to: (i) record somewhere that a clash occurred
+  between pages @pi@ and @pj@, and (ii) store the full key @maxKey pi@
+  separately. The record of clashes can be implemented as simply as a single bit
+  per page. Note that the first page's bit is always going to be @False@. Then,
+  because we are storing full keys only for some pages, we store the full keys
+  in a 'Map'.
+
+  The example below shows a simplified view of the compact index implementation
+  so far. As an example, we store the @32@ most significant bits of each minimum
+  key in the @primary@ index, the record of clashes is called @clashes@, and the
+  @IntMap@ is named the @tieBreaker@ map. @search3@ can at any point during the
+  search, consult @clashes@ and @tieBreaker@ to /break ties/.
+
+  > --              (primary         , clashes      , tieBreaker)
+  > type Index3 k = (VU.Vector Word32, VU.Vector Bit, IntMap k  )
+  >
+  > mkIndex3 :: Run k v -> Index3
+  > mkIndex3 = -- elided
+  >
+  > search3 :: k -> Index3 -> pageNo
+  > search3 = -- elided
+
+  Let's compare the memory size of @Index2@ with @Index3@. Say we have @O(n)@
+  pages, and keys that are @k@ bits each, then the memory size of @Index2@ is
+  @O(nk)@ bits. Say we have @O(n)@ pages and we store only the @m@ most
+  significant bits, then the memory size of @Index3@ is @O(n m
+  + 2 n + E[collision n m] * k)@ bits, where the last summand is the expected
+  number of collisions for @n@ independently generated hashes of bit-size @m@.
+  So, @Index3@ is not strictly an improvement over @Index2@ if we look at memory
+  complexity alone. However, /in practice/ @Index3@ is an improvement over
+  @Index2@ if we can pick an @m@ that is (i) much smaller than @k@ and (ii)
+  keeps @E[collision n m]@ small (if not close to @0@).
+
+  For example, storing the first @48@ bits of @2.5@ million SHA256 hashes
+  reduces memory size from @n * 256@ bits to @n * 48@ bits, because the expected
+  number of collisions is smaller than @1@.
+
+  TODO: the expected number of collisions is (n^2 - n ) / (2 * m) for hashing
+  @n@ distinct values into @m@ distinct buckets. Is the following reasoning
+  correct?
```

Comment:
I think the result is correct, but I'm not sure the reasoning is correct.

I'm not sure you can split an O(x * y) into an O(x) * O(y).

But I think you can reduce from O(n^2 + n) to O(n^2). The rule of thumb is the highest power in a polynomial dominates.

Reply:
I'll remove the derivation, but I'll keep the result